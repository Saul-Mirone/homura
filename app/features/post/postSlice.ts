import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { channel } from '../../channel/child';
import type { AppThunk, RootState } from '../../store';

type State = {
  content: string;
};

const initialState: State = {
  content: '',
};

const postSlice = createSlice({
  name: 'post',
  initialState,
  reducers: {
    setContent: (state, action: PayloadAction<string>) => {
      state.content = action.payload;
    },
  },
});

export const { setContent } = postSlice.actions;

export const getPostContentById = (postId: number): AppThunk => async (
  dispatch
) => {
  const post = await channel.getPostById(postId);
  dispatch(setContent(post.content));
};

export const postReducer = postSlice.reducer;

export const selectPost = (state: RootState) => {
  const { activeId, posts } = state.list;
  const { content } = state.post;

  if (!activeId)
    return {
      activeId,
    };

  const target = posts.find((x) => x.id === activeId);

  if (!target)
    return {
      activeId,
    };

  const { title, sourceName, unread, starred, date } = target;

  const post = {
    title,
    sourceName,
    unread,
    date,
    starred,
    content,
  };

  return {
    activeId,
    post,
  };
};
