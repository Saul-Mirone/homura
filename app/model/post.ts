import {
  Association,
  BelongsToCreateAssociationMixin,
  BelongsToGetAssociationMixin,
  BelongsToSetAssociationMixin,
  DataTypes,
  Model,
  Optional,
  Sequelize,
} from 'sequelize';
import type { Source } from './source';

export interface PostAttributes {
  id: number;
  sourceId: number;
  guid: string;
  title: string;
  content: string;
  unread: boolean;
  starred: boolean;
  date: Date;
}

export type CreatePostAttributes = Optional<
  PostAttributes,
  'id' | 'starred' | 'unread'
>;

export class Post
  extends Model<PostAttributes, CreatePostAttributes>
  implements PostAttributes {
  public id!: number;

  public sourceId!: number;

  public guid!: string;

  public title!: string;

  public content!: string;

  public unread!: boolean;

  public starred!: boolean;

  public date!: Date;

  public toJSON!: () => PostAttributes;

  public getSource!: BelongsToGetAssociationMixin<Source>;

  public setSource!: BelongsToSetAssociationMixin<Source, number>;

  public createSource!: BelongsToCreateAssociationMixin<Source>;

  public static associations: {
    source: Association<Post, Source>;
  };
}

export const initPost = (sequelize: Sequelize): void => {
  Post.init(
    {
      id: {
        type: DataTypes.INTEGER,
        autoIncrement: true,
        primaryKey: true,
      },
      sourceId: {
        type: DataTypes.INTEGER,
        allowNull: false,
      },
      guid: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      title: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      content: {
        type: DataTypes.TEXT,
        allowNull: false,
      },
      unread: {
        type: DataTypes.BOOLEAN,
        defaultValue: true,
      },
      starred: {
        type: DataTypes.BOOLEAN,
        defaultValue: false,
      },
      date: {
        type: DataTypes.DATE,
        allowNull: false,
      },
    },
    {
      sequelize,
      tableName: 'posts',
    }
  );
};
