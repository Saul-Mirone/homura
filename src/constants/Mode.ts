export enum Mode {
    All = 'All',
    Unread = 'Unread',
    Starred = 'Starred',
}
