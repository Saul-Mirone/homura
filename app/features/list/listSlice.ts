import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { DateTime } from 'luxon';
import { channel } from '../../channel/child';
import { Preset } from '../../constants/Preset';
import type { AppThunk, RootState } from '../../store';

type PostItem = {
  id: number;
  sourceId: number;
  title: string;
  sourceName: string;
  unread: boolean;
  starred: boolean;
  date: string;
  icon: string | null;
};

type TimeGroup = {
  time: string;
  posts: PostItem[];
};

type State = {
  groups: TimeGroup[];
  activeId: number | undefined;
};

const initialState: State = {
  groups: [],
  activeId: undefined,
};

type ISOString = string;

const listSlice = createSlice({
  name: 'list',
  initialState,
  reducers: {
    loadAll: (
      state,
      action: PayloadAction<Array<PostItem & { date: ISOString }>>
    ) => {
      state.groups = [...action.payload]
        .map(({ date, ...x }) => ({
          ...x,
          date: DateTime.fromISO(date),
        }))
        .sort((x, y) => y.date.toSeconds() - x.date.toSeconds())
        .map(({ date, ...x }) => ({
          ...x,
          date: date.toFormat('LLL dd, yyyy'),
        }))
        .reduce((acc, cur) => {
          const result = acc.findIndex((x) => x.time === cur.date);

          if (result < 0) {
            return [
              ...acc,
              {
                time: cur.date,
                posts: [cur],
              },
            ];
          }

          const target = acc[result];

          return [
            ...acc.slice(0, result),
            {
              ...target,
              posts: [...target.posts, cur],
            },
            ...acc.slice(result + 1),
          ];
        }, [] as TimeGroup[]);
    },
    setActiveId: (state, action: PayloadAction<number>) => {
      state.activeId = action.payload;
    },
    markActiveUnread: (state, action: PayloadAction<boolean>) => {
      const target = state.groups
        .flatMap((x) => x.posts)
        .find((x) => x.id === state.activeId);

      if (!target) return;

      target.unread = action.payload;
    },
  },
});

export const { loadAll, setActiveId, markActiveUnread } = listSlice.actions;

export const loadListBySourceId = (
  sourceId: number | Preset
): AppThunk => async (dispatch) => {
  if (typeof sourceId !== 'number') return;
  const source = await channel.getSourceById(sourceId);

  const posts = source.posts.map(
    ({ id, title, unread, date, starred, sourceId: postSourceId }) => ({
      id,
      sourceId: postSourceId,
      title,
      sourceName: source.name,
      icon: source.icon,
      date: date.toISOString(),
      unread,
      starred,
    })
  );

  dispatch(loadAll(posts));
};

export const markActiveUnreadAs = (unread: boolean): AppThunk => async (
  dispatch,
  getState
) => {
  const state = getState();
  if (!state.list.activeId) return;
  await channel.setPostUnread(state.list.activeId, unread);
  dispatch(markActiveUnread(unread));
};

export const listReducer = listSlice.reducer;

export const selectList = (state: RootState) => ({
  ...state.list,
  activeSourceId: state.source.activeId,
  mode: state.mode,
});
