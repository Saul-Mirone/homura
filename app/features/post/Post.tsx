import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Reader } from '../../components/Reader';
import { Toolkit } from '../../components/Reader/Toolkit';
import { AppDispatch } from '../../store';
import { markActiveStarredAs, markActiveUnreadAs } from '../list/listSlice';
import { getPostContentById, selectPost } from './postSlice';

export const Post: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const { post, activeId } = useSelector(selectPost);

  React.useEffect(() => {
    if (!activeId) return;
    dispatch(getPostContentById(activeId));
  }, [activeId, dispatch]);

  const toolkit = post ? (
    <Toolkit
      starred={post.starred}
      onSwitchStarred={(x) => dispatch(markActiveStarredAs(!x))}
      unread={post.unread}
      onSwitchUnread={(x) => dispatch(markActiveUnreadAs(!x))}
    />
  ) : null;

  return <Reader post={post} toolkit={toolkit} />;
};
