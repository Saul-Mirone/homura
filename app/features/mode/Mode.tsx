import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Toolkit } from '../../components/SubSideBar/ToolKit';
import { AppDispatch } from '../../store';
import { selectMode, switchMode } from './modeSlice';

export const Mode: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const mode = useSelector(selectMode);

  return <Toolkit mode={mode} onSwitchMode={(m) => dispatch(switchMode(m))} />;
};
