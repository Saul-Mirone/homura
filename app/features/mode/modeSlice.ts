import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { Mode } from '../../constants/Mode';
import type { RootState } from '../../store';

const modeSlice = createSlice({
  name: 'mode',
  initialState: Mode.All,
  reducers: {
    switchMode: (_, action: PayloadAction<Mode>) => action.payload,
  },
});

export const { switchMode } = modeSlice.actions;

export const modeReducer = modeSlice.reducer;

export const selectMode = (state: RootState) => state.mode;
