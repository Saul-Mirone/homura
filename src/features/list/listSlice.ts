import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { DateTime } from 'luxon';
import { channel } from '../../channel/child';
import { format } from '../../constants/Date';
import { Mode } from '../../constants/Mode';
import { Preset } from '../../constants/Preset';
import { Post, Source } from '../../model/types';
import type { AppThunk, RootState } from '../../store';
import { decCount, incCount, setCountToZero } from '../source/sourceSlice';

type PostKeys =
  | 'id'
  | 'sourceId'
  | 'title'
  | 'unread'
  | 'starred'
  | 'link'
  | 'date';

type PostItem = Pick<Post, PostKeys> &
  Pick<Source, 'icon'> & { sourceName: string };

type TimeGroup = {
  time: string;
  posts: PostItem[];
};

type State = {
  posts: PostItem[];
  activeId: number | undefined;
  filter: string;
};

const initialState: State = {
  posts: [],
  activeId: undefined,
  filter: '',
};

const listSlice = createSlice({
  name: 'list',
  initialState,
  reducers: {
    reset: (state) => {
      state.posts = [];
      state.activeId = undefined;
      state.filter = '';
    },
    loadAll: (state, action: PayloadAction<PostItem[]>) => {
      state.posts = action.payload;
    },
    setActiveId: (state, action: PayloadAction<number | undefined>) => {
      state.activeId = action.payload;
    },
    markActiveUnread: (state, action: PayloadAction<boolean>) => {
      const target = state.posts.find((x) => x.id === state.activeId);

      if (!target) return;

      target.unread = action.payload;
    },
    markActiveStarred: (state, action: PayloadAction<boolean>) => {
      const target = state.posts.find((x) => x.id === state.activeId);

      if (!target) return;

      target.starred = action.payload;
    },
    markAllRead: (state) => {
      state.posts.forEach((post) => {
        post.unread = false;
      });
    },
    setFilter: (state, action: PayloadAction<string>) => {
      state.filter = action.payload;
    },
  },
});

export const {
  loadAll,
  setActiveId,
  markActiveUnread,
  markActiveStarred,
  markAllRead,
  setFilter,
  reset,
} = listSlice.actions;

const loadBySourceId = async (sourceId: number, mode: Mode) => {
  const posts = await channel.getSourceById(sourceId);

  return posts
    .map(({ id, title, unread, date, starred, name, icon, link }) => ({
      id,
      sourceId,
      title,
      sourceName: name,
      icon,
      date,
      link,
      unread,
      starred,
    }))
    .filter((x) => {
      switch (mode) {
        case Mode.Starred:
          return x.starred;
        case Mode.Unread:
          return x.unread;
        case Mode.All:
        default:
          return true;
      }
    });
};
const loadByPreset = async (preset: Preset) => {
  return (await channel.getPostByPreset(preset)).map(
    ({ id, title, unread, date, starred, name, icon, sourceId, link }) => ({
      id,
      sourceId,
      title,
      sourceName: name,
      link,
      icon,
      unread,
      starred,
      date,
    })
  );
};

export const loadListBySourceId = (
  sourceIdOrPreset: number | Preset,
  mode: Mode
): AppThunk => async (dispatch) => {
  if (typeof sourceIdOrPreset === 'number') {
    const posts = await loadBySourceId(sourceIdOrPreset, mode);

    dispatch(loadAll(posts));
    return;
  }

  const posts = await loadByPreset(sourceIdOrPreset);
  dispatch(loadAll(posts));
};

export const markAllAsRead = (): AppThunk => async (dispatch, getState) => {
  const state = getState();
  const { activeId, list } = state.source;

  if (typeof activeId === 'number') {
    await channel.markAllAsReadBySourceId(activeId);

    setCountToZero(activeId);
  } else if (activeId && [Preset.Unread, Preset.All].includes(activeId)) {
    await Promise.all(list.map((x) => channel.markAllAsReadBySourceId(x.id)));
    list.forEach((x) => {
      setCountToZero(x.id);
    });
  }
  dispatch(markAllRead());
};

export const markActiveUnreadAs = (unread: boolean): AppThunk => async (
  dispatch,
  getState
) => {
  const state = getState();
  if (!state.list.activeId) return;
  await channel.setPostUnread(state.list.activeId, unread);
  const { posts } = state.list;
  const active = posts.find((x) => x.id === state.list.activeId);
  if (active && state.mode !== Mode.Starred) {
    if (unread) {
      dispatch(incCount(active.sourceId));
    } else {
      dispatch(decCount(active.sourceId));
    }
  }
  dispatch(markActiveUnread(unread));
};
export const markActiveStarredAs = (starred: boolean): AppThunk => async (
  dispatch,
  getState
) => {
  const state = getState();
  if (!state.list.activeId) return;
  await channel.setPostStarred(state.list.activeId, starred);
  const { posts } = state.list;
  const active = posts.find((x) => x.id === state.list.activeId);
  if (active && state.mode === Mode.Starred) {
    if (starred) {
      dispatch(incCount(active.sourceId));
    } else {
      dispatch(decCount(active.sourceId));
    }
  }
  dispatch(markActiveStarred(starred));
};

export const listReducer = listSlice.reducer;

export const selectList = (state: RootState) => {
  const { posts, activeId, filter } = state.list;
  const groups = [...posts]
    .map(({ date, ...x }) => ({
      ...x,
      date: DateTime.fromISO(date),
    }))
    .sort((x, y) => y.date.toSeconds() - x.date.toSeconds())
    .map(({ date, ...x }) => ({
      ...x,
      date: date.toFormat(format),
    }))
    .filter((x) => x.title.toLowerCase().includes(filter.toLowerCase()))
    .reduce((acc, cur) => {
      const result = acc.findIndex((x) => x.time === cur.date);

      if (result < 0) {
        return [
          ...acc,
          {
            time: cur.date,
            posts: [cur],
          },
        ];
      }

      const target = acc[result];

      return [
        ...acc.slice(0, result),
        {
          ...target,
          posts: [...target.posts, cur],
        },
        ...acc.slice(result + 1),
      ];
    }, [] as any[]) as TimeGroup[];
  return {
    activeId,
    groups,
    activeSourceId: state.source.activeId,
    mode: state.mode,
  };
};
