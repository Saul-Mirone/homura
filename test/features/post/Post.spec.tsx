/* eslint-disable @typescript-eslint/no-explicit-any */

// organize-imports-ignore
// make sure mock file on the top of imports
import { mockChannel } from '../../test-tools/mockChannel';

import { configureStore } from '@reduxjs/toolkit';
import { fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import * as listSlice from '../../../src/features/list/listSlice';
import { Post } from '../../../src/features/post/Post';
import * as postSlice from '../../../src/features/post/postSlice';
import * as modeSlice from '../../../src/features/mode/modeSlice';
import { Mode } from '../../../src/constants/Mode';
import { post } from '../../fixture/post';

function setup(
    preloadedState: {
        list: Partial<listSlice.State>;
        post: postSlice.State;
        mode: Mode;
    } = {
        list: {},
        post: {
            content: '',
        },
        mode: Mode.All,
    },
) {
    const initialListState = {
        posts: post,
        activeId: undefined,
        filter: '',
    };
    const state: {
        list: listSlice.State;
        post: postSlice.State;
        mode: Mode;
    } = {
        list: { ...initialListState, ...preloadedState.list },
        post: preloadedState.post,
        mode: preloadedState.mode,
    };
    const store = configureStore({
        reducer: { list: listSlice.listReducer, post: postSlice.postReducer, mode: modeSlice.modeReducer },
        preloadedState: state,
    });

    const utils = render(
        <Provider store={store}>
            <Post />
        </Provider>,
    );

    return {
        ...utils,
        store,
        setActiveId: (id: number) => store.dispatch(listSlice.setActiveId(id)),
        setMode: (mode: Mode) => store.dispatch(modeSlice.switchMode(mode)),
        el: {
            get post() {
                return utils.getByRole('article');
            },
            get toolkit() {
                return utils.getByRole('group');
            },
            get starButton() {
                return utils.getByLabelText('Mark as starred');
            },
            get unreadButton() {
                return utils.getByLabelText('Mark as unread');
            },
            get shareButton() {
                return utils.getByLabelText('Open in browser');
            },
        },
    };
}

test('Post', async () => {
    const mockScrollTo = jest.fn();
    HTMLElement.prototype.scrollTo = mockScrollTo;

    const mockOpen = jest.fn();
    global.open = mockOpen;

    mockChannel.setPostStarred.mockResolvedValue(undefined as any);
    mockChannel.setPostUnread.mockResolvedValue(undefined as any);
    mockChannel.getPostById.mockResolvedValue({ content: 'fake-content' } as any);

    const getPostContentByIdSpy = jest.spyOn(postSlice, 'getPostContentById');
    const markAsStarredSpy = jest.spyOn(listSlice, 'markAsStarred');
    const markActiveUnreadAsSpy = jest.spyOn(listSlice, 'markAsUnread');

    const { el, setActiveId, getByAltText } = setup();

    expect(getByAltText('homura')).toHaveAttribute('src');
    expect(el.post).toMatchSnapshot();

    setActiveId(20);
    expect(el.post).toMatchSnapshot();

    setActiveId(1);

    await waitFor(() => expect(getPostContentByIdSpy).toBeCalledWith(1));

    expect(mockScrollTo).toBeCalledWith(0, 0);

    expect(el.post).toMatchSnapshot();

    fireEvent.click(el.starButton);
    await waitFor(() => expect(markAsStarredSpy).toBeCalledTimes(1));
    expect(mockChannel.setPostStarred).toBeCalledWith(1, false);

    fireEvent.click(el.unreadButton);
    await waitFor(() => expect(markActiveUnreadAsSpy).toBeCalledTimes(1));
    expect(mockChannel.setPostUnread).toBeCalledWith(1, true);

    fireEvent.click(el.shareButton);
    expect(mockOpen).toBeCalledWith('test-link-1');
});
