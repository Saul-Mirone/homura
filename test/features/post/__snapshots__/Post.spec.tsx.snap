// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post 1`] = `
<div
  class="reader-container"
  role="article"
>
  <div
    class="reader-empty"
  >
    <div>
      <img
        alt="homura"
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;

exports[`Post 2`] = `
<div
  class="reader-container"
  role="article"
>
  <div
    class="reader-empty"
  >
    <div>
      <img
        alt="homura"
        src="test-file-stub"
      />
    </div>
  </div>
</div>
`;

exports[`Post 3`] = `
<div
  class="reader-container"
  role="article"
>
  <div
    class="flex justify-end"
    role="group"
  >
    <div
      class="inline-flex my-1 mr-1 text-gray-800"
    >
      <div
        aria-label="Mark as starred"
        class="p-2  py-1"
        role="button"
        tabindex="0"
      >
        <div
          class="w-5 h-5 m-auto  "
        >
          <svg
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.049 2.927c.3-.921 1.603-.921 1.902 0l1.07 3.292a1 1 0 00.95.69h3.462c.969 0 1.371 1.24.588 1.81l-2.8 2.034a1 1 0 00-.364 1.118l1.07 3.292c.3.921-.755 1.688-1.54 1.118l-2.8-2.034a1 1 0 00-1.175 0l-2.8 2.034c-.784.57-1.838-.197-1.539-1.118l1.07-3.292a1 1 0 00-.364-1.118L2.98 8.72c-.783-.57-.38-1.81.588-1.81h3.461a1 1 0 00.951-.69l1.07-3.292z"
            />
          </svg>
        </div>
      </div>
      <div
        aria-label="Mark as unread"
        class="p-2  py-1"
        role="button"
        tabindex="0"
      >
        <div
          class="w-5 h-5 m-auto  "
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 8l7.89 5.26a2 2 0 002.22 0L21 8M5 19h14a2 2 0 002-2V7a2 2 0 00-2-2H5a2 2 0 00-2 2v10a2 2 0 002 2z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
      <div
        aria-label="Open in browser"
        class="p-2  py-1"
        role="button"
        tabindex="0"
      >
        <div
          class="w-5 h-5 m-auto  "
        >
          <svg
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.684 13.342C8.886 12.938 9 12.482 9 12c0-.482-.114-.938-.316-1.342m0 2.684a3 3 0 110-2.684m0 2.684l6.632 3.316m-6.632-6l6.632-3.316m0 0a3 3 0 105.367-2.684 3 3 0 00-5.367 2.684zm0 9.316a3 3 0 105.368 2.684 3 3 0 00-5.368-2.684z"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  <div
    class="thin-scroll"
  >
    <div
      class="reader-post"
    >
      <hgroup>
        <small>
          Invalid DateTime
        </small>
        <div>
          test-source-name-1
        </div>
        <h1>
          test-title-1
        </h1>
      </hgroup>
      <main>
        fake-content
      </main>
    </div>
  </div>
</div>
`;
